ctxutil: Add `ctxutil.WhenDone`

miretskiy opened the pull request:
Add `ctxutil.WhenDone` function which invokes a callback when parent context becomes done.

```
parent, cancelParent := context.WithTimeout(....)
ctxutil.WhenDone(parent, func() {....})
```

Normally, in order to invoke a function when context becomes done, a go routine should be spun up that listens on `ctx.Done()` channel.  Sometimes, this is too heavy weight. `ctxutil.WhenDone` adds such a mechanism, which under most circumstances, simply invokes the function when context completes.

When parent context becomes done, the function is invoked with the result (Err()) of the parent context.

If the parent context never completes, `WhenDone` returns an error.

Go versions 1.20 or higher also add ability to augment context error message with "cause" error.

Epic: None

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/103889)
<!-- Reviewable:end -->

Commit:
ctxutil: Fix context cancellation propagation in go 1.21

Update ctxutil package to correctly handle context cancellation
propagation for go 1.21.

Remove go 1.19 compatability.

Informs #112088.
Needed for #111928.

Release note: None
